File: src/main/java/com/github/jferard/classwriter/internal/context/MethodContext.java
/*
 * ClassWriter - A minimal Java bytecode writer. Creates classes, methods, interfaces...
 *     Copyright (C) 2018 J. Férard <https://github.com/jferard>
 *
 * This file is part of ClassWriter.
 *
 * ClassWriter is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * ClassWriter is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.jferard.classwriter.internal.context;

import com.github.jferard.classwriter.api.Label;
import com.github.jferard.classwriter.api.PlainClassRef;
import com.github.jferard.classwriter.internal.attribute.PositionAndLineNumber;
import com.github.jferard.classwriter.internal.attribute.stackmap.FrameContext;
import com.github.jferard.classwriter.internal.attribute.stackmap.StackMapFrame;
import com.github.jferard.classwriter.internal.attribute.stackmap.StackMapFrameData;
import com.github.jferard.classwriter.internal.attribute.stackmap.VerificationType;
import com.github.jferard.classwriter.internal.instruction.ExceptionInCode;
import com.github.jferard.classwriter.internal.instruction.ExceptionInCodeBuilder;
import com.github.jferard.classwriter.internal.instruction.Instruction;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class MethodContext {
    public static MethodContext create(int nargs) {
        return new MethodContext(0, FrameContext.create(), OffsetsContext.create(nargs),
                new ArrayList<>(), new Stack<>(), new ArrayList<>());
    }

    private final List<ExceptionInCode> exceptions;
    private OffsetsContext offsetsContext;
    private FrameContext frameContext;
    private int offset;
    private ExceptionInCodeBuilder curExceptionBuilder;
    private Stack<StackMapFrame> stackFrames;
    private Offsets finalOffsetsContext;

    public List<PositionAndLineNumber> getPositionsAndLineNumbers() {
        return this.positionsAndLineNumbers;
    }

    private List<PositionAndLineNumber> positionsAndLineNumbers;

    MethodContext(int offset, FrameContext frameContext, OffsetsContext offsetsContext,
                  final List<ExceptionInCode> exceptions, Stack<StackMapFrame> stackFrames,
                  List<PositionAndLineNumber> positionsAndLineNumbers) {
        this.offset = offset;
        this.frameContext = frameContext;
        this.offsetsContext = offsetsContext;
        this.exceptions = exceptions;
        this.stackFrames = stackFrames;
        this.positionsAndLineNumbers = positionsAndLineNumbers;
    }

    /**
     * Add a subroutine
     *
     * @param instruction the subroutine
     */
    public void addSubroutine(Instruction instruction) {
        this.offsetsContext.addSubroutine(instruction);
    }

    /**
     * should store the stackframe
     */
    public void storeGoto(Label label) {
        this.offsetsContext.storeGoto(label);
    }

    public void storeLabel(Label label) {
        this.offsetsContext.storeLabel(label, this.frameContext.toData());
    }

    public void storeJsr(Instruction instruction) {
        this.offsetsContext.storeJsr(instruction);
    }


    public void storeBranch(Label label) {
        this.offsetsContext.storeBranch(label);
    }

    public void storePadding(int padding) {
        this.offsetsContext.storePadding(padding);
    }

    /**
     * Fix wide index goto and jsr
     * What about padding (lookupswitch, tableswitch) ?
     */
    public void normalize(GlobalContext context) {
        this.finalOffsetsContext = this.offsetsContext.normalize(context, this);
    }

    public int getLabelOffset(Label label) {
        return this.finalOffsetsContext.getLabelOffset(label);
    }

    public long getSubroutineOffset(Instruction instruction) {
        return this.finalOffsetsContext.getSubroutineOffset(instruction);
    }


    public int getMaxStack() {
        return this.frameContext.getMaxStack();
    }

    public int getMaxLocals() {
        return this.frameContext.getMaxLocals();
    }

    public int getCurOffset() {
        return this.offset;
    }

    public void storeBeginException(PlainClassRef classRef, Label label, int curOffset) {
        this.curExceptionBuilder = ExceptionInCode.builder(classRef, label, curOffset);
    }

    public void storeEndException(int curOffset) {
        this.exceptions.add(this.curExceptionBuilder.endOffset(curOffset).build());
    }

    public List<ExceptionInCode> getExceptions() {
        return this.exceptions;
    }

    public MethodContext clone() {
        return new MethodContext(this.offset, FrameContext.create(), OffsetsContext.create(0),
                this.exceptions, new Stack<>(), this.positionsAndLineNumbers);
    }

    public void merge(MethodContext otherContext) {
        // merge frame context
        throw new IllegalStateException();
    }

    public void offsetDelta(int delta) {
        this.offset += delta;
    }

    public void stackEmpty() {
        this.frameContext.stackEmpty();
    }

    public void stackPush(VerificationType type) {
        this.frameContext.stackPush(type);
    }

    public void localsPop(int n) {
        this.frameContext.localsPop(n);
    }

    public void localsPush(VerificationType type) {
        this.frameContext.localsPush(type);
    }

    public VerificationType stackToTop() {
        return this.frameContext.stackToTop();
    }

    public VerificationType stackPop() {
        return this.frameContext.stackPop();
    }

    public VerificationType stackPopOneByte() {
        return this.frameContext.stackPopOneByte();
    }

    public VerificationType localsGet(int index) {
        return this.frameContext.localsGetVerificationType(index);
    }

    public StackMapFrameData getStackMapFrameData() {
        return this.frameContext.toData();
    }

    public void assertTypeEquals(VerificationType expectedType, VerificationType actualType) {
        if (expectedType.equals(actualType)) {
            return;
        }

        throw new IllegalStateException("" + expectedType + " vs " + actualType);
    }

    public void assertTypeAssignable(VerificationType expectedType, VerificationType actualType) {
        if (actualType.isAssignable(expectedType)) {
            return;
        }

        throw new IllegalStateException("" + expectedType + " vs " + actualType);
    }

    public ReferenceVerificationType assertIsArray(VerificationType verificationType) {
        if (verificationType instanceof ReferenceVerificationType) {
            final ReferenceVerificationType referenceVerificationType =
                    (ReferenceVerificationType) verificationType;
            if (referenceVerificationType.isArray()) {
                return referenceVerificationType;
            }
        }

        throw new IllegalStateException();
    }

    public VerificationType stackPeek() {
        return this.frameContext.stackPeek();
    }

    public void localsSet(int localIndex, VerificationType stackType) {
        this.frameContext.localsSet(localIndex, stackType);
    }

    public void stackClear() {
        this.frameContext.stackClear();
    }

    public void storeLineNumber(int lineNumber) {
        this.positionsAndLineNumbers.add(new PositionAndLineNumber(offset, lineNumber));
    }
}

File: src/main/kotlin/com/github/jferard/classwriter/internal/context/MethodContext.kt
/*
 * ClassWriter - A minimal JVM bytecode writer. Creates classes, methods, interfaces...
 *     Copyright (C) 2018 J. Férard <https://github.com/jferard>
 *
 * This file is part of ClassWriter.
 *
 * ClassWriter is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * ClassWriter is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.jferard.classwriter.internal.context

import com.github.jferard.classwriter.api.Label
import com.github.jferard.classwriter.api.PlainClassRef
import com.github.jferard.classwriter.internal.attribute.PositionAndLineNumber
import com.github.jferard.classwriter.internal.attribute.stackmap.FrameContext
import com.github.jferard.classwriter.internal.attribute.stackmap.StackMapFrame
import com.github.jferard.classwriter.internal.attribute.stackmap.StackMapFrameData
import com.github.jferard.classwriter.internal.attribute.stackmap.VerificationType
import com.github.jferard.classwriter.api.instruction.ExceptionInCode
import com.github.jferard.classwriter.api.instruction.ExceptionInCodeBuilder
import com.github.jferard.classwriter.api.instruction.Instruction
import java.util.*
import kotlin.collections.ArrayList


class MethodContext(var curOffset: Int, private val frameContext: FrameContext,
                                         private val offsetsContext: OffsetsContext,
                                         private val exceptions: MutableList<ExceptionInCode>,
                                         private val stackFrames: Stack<StackMapFrame>,
                                         private val positionsAndLineNumbers: MutableList<PositionAndLineNumber>) {
    private var curExceptionBuilder: ExceptionInCodeBuilder? = null
    private var finalOffsetsContext: Offsets? = null
    
    fun getPositionsAndLineNumbers(): List<PositionAndLineNumber> {
        return positionsAndLineNumbers
    }

    /**
     * Add a subroutine
     *
     * @param instruction the subroutine
     */
    fun addSubroutine(
            instruction: Instruction) {
        offsetsContext.addSubroutine(instruction)
    }

    /**
     * should store the stackframe
     */
    fun storeGoto(label: Label) {
        offsetsContext.storeGoto(label)
    }

    fun storeLabel(label: Label) {
        offsetsContext.storeLabel(label, frameContext.toData())
    }

    fun storeJsr(
            instruction: Instruction) {
        offsetsContext.storeJsr(instruction)
    }

    fun storeBranch(label: Label) {
        offsetsContext.storeBranch(label)
    }

    fun storePadding(padding: Int) {
        offsetsContext.storePadding(padding)
    }

    /**
     * Fix wide index goto and jsr
     * What about padding (lookupswitch, tableswitch) 
     */
    fun normalize(context: GlobalContext) {
        finalOffsetsContext = offsetsContext.normalize(context, this)
    }

    fun getLabelOffset(label: Label): Int {
        return finalOffsetsContext!!.getLabelOffset(label)
    }

    fun getSubroutineOffset(
            instruction: Instruction): Long {
        return finalOffsetsContext!!.getSubroutineOffset(instruction)
    }

    val maxStack: Int
        get() = frameContext.maxStack

    val maxLocals: Int
        get() = frameContext.maxLocals

    fun storeBeginException(classRef: PlainClassRef,
                            label: Label,
                            curOffset: Int) {
        curExceptionBuilder = ExceptionInCode.builder(classRef, label, curOffset)
    }

    fun storeEndException(curOffset: Int) {
        exceptions.add(curExceptionBuilder!!.endOffset(curOffset).build())
    }

    fun getExceptions(): List<ExceptionInCode> {
        return exceptions
    }

    fun clone(): MethodContext {
        return MethodContext(curOffset, FrameContext.create(),
                OffsetsContext.create(0),
                exceptions, Stack(), positionsAndLineNumbers)
    }

    fun merge(otherContext: MethodContext) { // merge frame context
        throw IllegalStateException()
    }

    fun offsetDelta(delta: Int) {
        curOffset += delta
    }

    fun stackEmpty() {
        frameContext.stackEmpty()
    }

    fun stackPush(type: VerificationType) {
        frameContext.stackPush(type)
    }

    fun localsPop(n: Int) {
        frameContext.localsPop(n)
    }

    fun localsPush(type: VerificationType) {
        frameContext.localsPush(type)
    }

    fun stackToTop(): VerificationType {
        return frameContext.stackToTop()
    }

    fun stackPop(): VerificationType {
        return frameContext.stackPop()
    }

    fun stackPopOneByte(): VerificationType {
        return frameContext.stackPopOneByte()
    }

    fun localsGet(index: Int): VerificationType {
        return frameContext.localsGetVerificationType(index)
    }

    val stackMapFrameData: StackMapFrameData
        get() = frameContext.toData()

    fun assertTypeEquals(expectedType: VerificationType,
                         actualType: VerificationType) {
        if (expectedType == actualType) {
            return
        }
        throw IllegalStateException("$expectedType vs $actualType")
    }

    fun assertTypeAssignable(expectedType: VerificationType,
                             actualType: VerificationType) {
        if (actualType.isAssignable(expectedType)) {
            return
        }
        throw IllegalStateException("$expectedType vs $actualType")
    }

    fun assertIsArray(verificationType: VerificationType): ReferenceVerificationType {
        if (verificationType is ReferenceVerificationType) {
            val referenceVerificationType: ReferenceVerificationType =
                    verificationType
            if (referenceVerificationType.isArray) {
                return referenceVerificationType
            }
        }
        throw IllegalStateException()
    }

    fun stackPeek(): VerificationType {
        return frameContext.stackPeek()
    }

    fun localsSet(localIndex: Int, stackType: VerificationType) {
        frameContext.localsSet(localIndex, stackType)
    }

    fun stackClear() {
        frameContext.stackClear()
    }

    fun storeLineNumber(lineNumber: Int) {
        positionsAndLineNumbers.add(PositionAndLineNumber(curOffset, lineNumber))
    }

    companion object {
        fun create(nargs: Int): MethodContext {
            return MethodContext(0, FrameContext.create(),
                    OffsetsContext.create(nargs),
                    ArrayList(), Stack(),
                    ArrayList())
        }
    }
}